qp_shogun/sortmerna/utils.py

```diff
@@ -10,11 +10,44 @@
 # This file contains functions used by multiple commands
 # -----------------------------------------------------------------------------
 
+import gzip
 from os.path import join, exists
 from functools import partial
 from qiita_client import ArtifactInfo
 
 
+# Returns a file list
+def _gzip_uncompress(input_filelist):
+    input_filelist_uncompressed = []
+    for input_file in input_filelist:
+        if input_file.endswith('.fastq.gz'):
+            filename_uncompressed = input_file.replace(
+                '.fastq.gz', '.fastq')
+            # Uncompressing and reading the .gz file
+            input = gzip.GzipFile(input_file, 'rb')
+            s = input.read()
+            input.close()
+            # Writing uncompressed data into a file
+            output = open(filename_uncompressed, 'wb')
+            output.write(s)
+            output.close()
+            # Return the .fastq filename
+            input_filelist_uncompressed.append(filename_uncompressed)
+        else:
+            raise ValueError('File %s has an unexpected name' % input_file)
+    return input_filelist_uncompressed
+
+
+def _gzip_compress(input_filename):
+    f_in = open(input_filename, 'rb')
+    output_filename = input_filename + '.gz'
+    f_out = gzip.open(output_filename, 'wb')
+    f_out.writelines(f_in)
+    f_out.close()
+    f_in.close()
+    return output_filename
+
```

Comment:
This looks fine, except it might be better to do a command-call to [pigz](https://zlib.net/pigz/) for parallel compression/decompression. It could use the same number of threads as would be used for sortmerna. Any thoughts? Otherwise this could be very slow for large files.

If we don't want to use pigz, then we might want to read and write data in chunks i.e. passing the number of bytes to the read method. I don't think it would be good to read the whole file into memory.

Reply:
Modified code to calf pigz for compression/decompression within the genrate_smr command